extensions/s3/src/main/java/io/deephaven/extensions/s3/S3OutputStream.java

```diff
@@ -0,0 +1,268 @@
+//
+// Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+//
+package io.deephaven.extensions.s3;
+
+import org.jetbrains.annotations.NotNull;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.S3Uri;
+import software.amazon.awssdk.services.s3.internal.multipart.SdkPojoConversionUtils;
+import software.amazon.awssdk.services.s3.model.AbortMultipartUploadRequest;
+import software.amazon.awssdk.services.s3.model.CompleteMultipartUploadRequest;
+import software.amazon.awssdk.services.s3.model.CompletedMultipartUpload;
+import software.amazon.awssdk.services.s3.model.CompletedPart;
+import software.amazon.awssdk.services.s3.model.CreateMultipartUploadRequest;
+import software.amazon.awssdk.services.s3.model.CreateMultipartUploadResponse;
+import software.amazon.awssdk.services.s3.model.UploadPartRequest;
+import software.amazon.awssdk.services.s3.model.UploadPartResponse;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import static io.deephaven.extensions.s3.S3ChannelContext.handleS3Exception;
+
+class S3OutputStream extends OutputStream {
+
+    /**
+     * @see <a href="https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html">Amazon S3 User Guide</a>
+     */
+    private static final int MIN_PART_NUMBER = 1;
+    private static final int MAX_PART_NUMBER = 10000;
+    private static final int INVALID_PART_NUMBER = -1;
+
+    private final S3Uri uri;
+    private final S3AsyncClient s3AsyncClient;
+    private final S3Instructions s3Instructions;
+
+    private final int partSize;
+    private final int numConcurrentParts; // TODO Better name for this
+
+    private final List<CompletedPart> completedParts;
+    private final List<OutgoingRequest> pendingRequests;
+
+    private int nextPartNumber;
+    private String uploadId;
+
+    S3OutputStream(
+            @NotNull final URI uri,
+            @NotNull final S3AsyncClient s3AsyncClient,
+            @NotNull final S3Instructions s3Instructions) {
+        this.uri = s3AsyncClient.utilities().parseUri(uri);
+        this.s3AsyncClient = s3AsyncClient;
+        this.s3Instructions = s3Instructions;
+
+        this.partSize = s3Instructions.partSize();
+        this.numConcurrentParts = s3Instructions.numConcurrentParts();
+        this.pendingRequests = new ArrayList<>(numConcurrentParts);
+
+        this.nextPartNumber = MIN_PART_NUMBER;
+        this.completedParts = new ArrayList<>();
+    }
+
+    public void write(int b) throws IOException {
+        write(new byte[] {(byte) b}, 0, 1);
+    }
```

Comment:
I kind of hate making an array for this. But maybe we're never doing single-byte writes anyway?